kafka/main.go

```diff
@@ -67,94 +87,154 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
-								providedServer,
-								logger,
-							)
-							if err != nil {
-								logger.Error().Err(err).Msg("Processing failed.")
-							}
-						}(msg, topicMapping, providedServer, logger)
-
-						if err := consumer.CommitMessages(ctx, msg); err != nil {
-							logger.Error().Msgf("Failed to commit: %s", err)
-						}
-					}
-				}
-			}
+// processMessagesIntoBatchPipeline fetches messages from Kafka indefinitely, pushes a
+// MessageContext into the batchPipeline to maintain message order, and then spawns a
+// goroutine that will process the message and push to errorResult of the MessageContext
+// when the processing completes. In case of an error, we panic from this function,
+// triggering the deferral which closes the batchPipeline channel. This will result in
+// readAndCommitBatchPipelineResults returning an error and the processing loop being recreated.
+func processMessagesIntoBatchPipeline(
+	ctx context.Context,
+	topicMappings []TopicMapping,
+	providedServer *server.Server,
+	reader *kafka.Reader,
+	batchPipeline chan *MessageContext,
+	logger *zerolog.Logger,
+) {
+	// During normal operation processMessagesIntoBatchPipeline will never complete and
+	// this deferral should not run. It's only called if we encounter some unrecoverable
+	// error.
+	defer func() {
+		close(batchPipeline)
+	}()
 
-			// The below block will close the producer connection when the error threshold is reached.
-			// @TODO: Test to determine if this Close() impacts the other goroutines that were passed
-			// the same topicMappings before re-enabling this block.
-			//for _, topicMapping := range topicMappings {
-			//	logger.Trace().Msg(fmt.Sprintf("Closing producer connection %v", topicMapping))
-			//	if err := topicMapping.ResultProducer.Close(); err != nil {
-			//		logger.Error().Msg(fmt.Sprintf("Failed to close writer: %e", err))
-			//	}
-			//}
-		}(topicMappings)
+	for {
+		msg, err := reader.FetchMessage(ctx)
+		if err != nil {
+			// Indicates batch has no more messages. End the loop for
+			// this batch and fetch another.
+			if err == io.EOF {
+				logger.Info().Msg("Batch complete")
+			} else if strings.ToLower(err.Error()) != "context deadline exceeded" {
+				logger.Error().Err(err).Msg("batch item error")
+				panic("failed to fetch kafka messages and closed channel")
+			}
+			// There are other possible errors, but the underlying consumer
+			// group handler handle retryable failures well. If further
+			// investigation is needed you can review the handler here:
+			// https://github.com/segmentio/kafka-go/blob/main/consumergroup.go#L729
+			continue
+		}
+		msgCtx := &MessageContext{
+			errorResult: make(chan error),
+			msg:         msg,
+		}
+		// If batchPipeline has been closed by an error in readAndCommitBatchPipelineResults,
+		// this write will panic, which is desired behavior, as the rest of the context
+		// will also have died and will be restarted from kafka/main.go
+		batchPipeline <- msgCtx
```

Comment:
Is the intention with the multiple channels to sync the go routines to keep order of the offsets e.g. multiple go routines read from same partition and complete at different times hence may commit offsets out of order and the channels are so keep that order?

Reply:
Does the batchPipeline need a buffer so it wont block and each message can then be processed concurrently. In `readAndCommitBatchPipelineResults` it will block when there is no result `msgCtx.errorResult` for a slow handler but if something is already on the channel it will process immediately.

Reply:
great catch. fixed